### ISSUE TITLE
lz4_compress_fast() and lz4_compress_hc()

### ISSUE THREAD

**Issue by IshaanKudchadkar** [lz4/lz4#1504]
I'm confused as to which cli command calls which internal function while executing. Does using a cli command "lz4 -1" use lz4_compress_fast() and "lz4 -9" use lz4_compress_HC()? Or how is it?

**Comment by t-mat** [comment]
> Does using a cli command "lz4 -1" use lz4_compress_fast()  and "lz4 -9" use lz4_compress_HC()?

Yes.  Actually, the `lz4` CLI doesn't use the public function directly.  But basically it uses equivalent function / algorithm.

----

As of lz4 v1.10.0, relation with CLI compression arguments and internal functions are:

- `-0` and `-1` are identical and they initialize the stream via [`LZ4_initStream()`](https://github.com/lz4/lz4/blob/v1.10.0/lib/lz4frame.c#L716).
- `-2` to `-12` calls [`LZ4_initStreamHC()`](https://github.com/lz4/lz4/blob/v1.10.0/lib/lz4frame.c#L720) with `compressionLevel`.
  - [`2 == LZ4HC_CLEVEL_MIN`](https://github.com/lz4/lz4/blob/v1.10.0/lib/lz4hc.h#L47)
  - [`12 == LZ4HC_CLEVEL_MAX`](https://github.com/lz4/lz4/blob/v1.10.0/lib/lz4hc.h#L50)

You can check them by placing breakpoint or `printf()` to [`LZ4F_compressBegin_internal()`](https://github.com/lz4/lz4/blob/v1.10.0/lib/lz4frame.c#L690).

----

@Cyan4973 ,  [`usage_longhelp()`](https://github.com/lz4/lz4/blob/v1.10.0/programs/lz4cli.c#L207-L210) shows the following message, but it seems it's wrong.

```
$ lz4 -H
...

Compression levels :
---------------------
-0 ... -2  => Fast compression, all identical
-3 ... -12 => High compression; higher number == more compression but slower
```

I think using `LZ4HC_CLEVEL_MIN` would be nicer to display the actual condition.

**Comment by IshaanKudchadkar** [comment]
So the lz4 CLI doesn't really use the above 2 functions I mentioned? If data was fed to the lz4 CLI then would it give similar stats (compression ratio and speeds) as compared to those 2 functions?

**Comment by Cyan4973** [comment]
> @Cyan4973 , [`usage_longhelp()`](https://github.com/lz4/lz4/blob/v1.10.0/programs/lz4cli.c#L207-L210) shows the following message, but it seems it's wrong.

You're right @t-mat , this should be fixed

**Comment by t-mat** [comment]
> So the lz4 CLI doesn't really use the above 2 functions I mentioned?

`lz4` CLI uses them when [`-l` (legacy)](https://github.com/lz4/lz4/blob/v1.10.0/programs/lz4.1.md?plain=1#L209) modifier argument is provided.

- `-l1` (legacy format, fast) [calls `LZ4_compress_fast()`](https://github.com/lz4/lz4/blob/v1.10.0/programs/lz4io.c#L789).  But it doesn't conform the LZ4 Frame Format.
- `-l9` (legacy format, high) [calls `LZ4_compress_HC()`](https://github.com/lz4/lz4/blob/v1.10.0/programs/lz4io.c#L808).  But it doesn't conform the LZ4 Frame Format.
- `-1` doesn't call `LZ4_compress_fast()`.  It uses LZ4/LZ4F streaming fnctions.
- `-9` doesn't call `LZ4_compress_HC()`.  It uses LZ4/LZ4F streaming functions.

> If data was fed to the lz4 CLI then would it give similar stats (compression ratio and speeds) as compared to those 2 functions?

They're similar but not identical/compatible.
